### ISSUE TITLE
[BUG] OpenSearchLoggerUsageTests not updated to reflect log4j2 functionality

### ISSUE THREAD

**Issue by Poojita-Raj** [opensearch-project/OpenSearch#1585]
**Describe the bug**
In OpenSearchLoggerUsageTests, the expected behaviour of the logger in certain tests is incorrect. Specifically for checkFailOrderOfExceptionArgument tests, which should pass.

**To Reproduce**
<img width="558" alt="image" src="https://user-images.githubusercontent.com/20815824/142516681-d7981ae9-e69f-48cb-b145-3ee88ccc3bbb.png">
The above line results in a failure as shown below.

**Expected behavior**
checkFailOrderOfExceptionArgument tests need to be modified to pass when an additional exception argument is passed in as the second argument.

**Screenshots**
<img width="929" alt="image" src="https://user-images.githubusercontent.com/20815824/142516578-eb7d52c4-c7c5-45a8-a23e-a479e5301267.png">

**Comment by andrross** [comment]
The following code produces the expected result of the substituting the parameter in the format string while still printing the exception stack trace:

```
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

public class LoggerTest {
    private static final Logger logger = LogManager.getLogger(LoggerTest.class);

    public static void main(String[] args) {
        logger.error("Hello, {}", "world", new RuntimeException("Test Exception"));
    }
}
```

```
16:46:15.577 [main] ERROR org.opensearch.LoggerTest - Hello, world
java.lang.RuntimeException: Test Exception
	at org.opensearch.LoggerTest.main(LoggerTest.java:18) [classes/:?]
```

The checks done by `OpenSearchLoggerUsageChecker` force developers to write the above log statement with a more verbose syntax:

```
logger.error(() -> new ParameterizedMessage("Hello, {}", "world"), new RuntimeException("Test Exception"));
```

I think we should relax the checks done by OpenSearchLoggerUsageChecker unless there's a good reason that I'm not aware of.

**Comment by Bukhtawar** [comment]
The support for exception parameterization(as the last one) was added in http://www.slf4j.org/faq.html#paramException. I think the checker needs to be reflective of this support

**Comment by minalsha** [comment]
@Poojita-Raj  see that you have a PR for addressing this issue and its closed without merging. Can you please look into this PR and issue and merge the PR to close the issue?

**Comment by minalsha** [comment]
@Poojita-Raj  can you share an update on this issue?

**Comment by minalsha** [comment]
@Poojita-Raj  can you please share an update here with whats pending for this issue?

**Comment by Poojita-Raj** [comment]
Hi, 

Issue is exception parameterization is not recognized by the OpenSearchLoggerUsageChecker. 

As you can see here:

https://github.com/opensearch-project/OpenSearch/blob/e455f56c749011d886e8939ae5bce0448be19f8c/test/logger-usage/src/main/java/org/opensearch/test/loggerusage/OpenSearchLoggerUsageChecker.java#L369-L374

The exception as in the case of : 

`logger.error("Hello, {}", "world", new RuntimeException("Test Exception"));`

is recognized as an object class instead of a throwable which then results in it wrongly deciding the number of arguments don't match up here: 

https://github.com/opensearch-project/OpenSearch/blob/e455f56c749011d886e8939ae5bce0448be19f8c/test/logger-usage/src/main/java/org/opensearch/test/loggerusage/OpenSearchLoggerUsageChecker.java#L392-L395

The next steps would be to: 

1. Look into why the argumentType in this case is not being recognized as a Throwable ( Ljava/lang/Throwable). 
2. This specific case is during one of the 5 types of method invocation (Ref: https://docs.oracle.com/javase/specs/jvms/se8/html/jvms-2.html#jvms-2.11.8 - invoke interface). We need to look into whether similar issues arise in other invocation cases.

**Comment by andrross** [comment]
> Look into why the argumentType in this case is not being recognized as a Throwable ( Ljava/lang/Throwable).

The logger checker in OpenSearch probably existed before the logging API would handle the case of an exception at the end of a list of parameters. I think we just need to update the logger checker to handle this case and not be overly strict.

**Comment by dbwiddis** [comment]
I spent a few hours looking into this.  I'm not sure this is doable with the log4j api.

> 1. Look into why the argumentType in this case is not being recognized as a Throwable ( Ljava/lang/Throwable).

Because the `argumentType` is coming from the method signature, not the actual subclass being tested.  Since we are using `log4j` the matching signature for `logger.info("{} failed too many times", FollowerChecker.this, exp);` is `void info(String message, Object p0, Object p1);`.   There is no signature with parameters String, Object, Throwable, nor can there be because it would be ambiguous.

The link @Bukhtawar provided in [this comment](https://github.com/opensearch-project/OpenSearch/issues/1585#issuecomment-974619528) confirms this observation, as it highlights that the exception printing behavior is internal, and that
> The java compiler will invoke the error method taking a String and two Object arguments

The only ways to get it to recognize the throwable are
 - the supplier syntax noted by @andrross in [this comment](https://github.com/opensearch-project/OpenSearch/issues/1585#issuecomment-973607464): `logger.info(() -> new ParameterizedMessage("{} failed too many times", FollowerChecker.this), exp);` That syntax matches `void info(MessageSupplier messageSupplier, Throwable throwable);`
 - a similar option to the above but without the `() ->` part: `logger.info(new ParameterizedMessage("{} failed too many times", FollowerChecker.this), exp);` which matches `void info(Message message, Throwable throwable);`.

I did a pretty thorough exploration of every value available in the parameters (`methodInsn` is the `info()` call being considered) and it simply doesn't have the information to identify that third parameter as a throwable.

Closest we can get is to "always allow an extra parameter" but [I can't see any (reasonable) way of knowing whether it's a throwable](https://xkcd.com/1425/).

I recommend closing this issue as wontfix/cantfix.

**Comment by andrross** [comment]
@dbwiddis I agree with your conclusion above with the tooling we have but do find it rather unsatisfying. My IDE is able to handle the case of the trailing throwable:

<img width="567" alt="image" src="https://github.com/opensearch-project/OpenSearch/assets/9160839/e079a032-91ab-45e7-a234-2487e0ce8b4b">

Surely there exists some tooling for us to get the same behavior at build time?
